File: neetcode/blind75_rust/src/code/implement_trie.rs
use std::collections::HashMap;

#[derive(PartialEq, Eq, Debug, Clone)]
struct Node {
    map: HashMap<char, Node>,
}

impl Node {
    fn new() -> Self {
        Node {
            map: HashMap::<char, Node>::new(),
        }
    }
}

struct Trie {
    root: Node,
}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl Trie {
    fn new() -> Self {
        Trie { root: Node::new() }
    }

    fn insert(&mut self, word: String) {
        let mut p = &mut self.root;

        for c in word.chars() {
            if !p.map.contains_key(&c) {
                p.map.insert(c, Node::new());
            }
            p = p.map.get_mut(&c).unwrap();
        }

        p.map.insert('&', Node::new());
    }

    fn search(&self, word: String) -> bool {
        match Trie::find(self, word) {
            Some(node) => {
                println!("{:?}", node);
                if node.map.contains_key(&'&') {
                    true
                } else {
                    false
                }
            }
            None => false,
        }
    }

    fn starts_with(&mut self, prefix: String) -> bool {
        match Trie::find(self, prefix) {
            Some(_) => true,
            None => false,
        }
    }

    fn find(&self, prefix: String) -> Option<Node> {
        let mut p = &self.root;
        for c in prefix.chars() {
            if !p.map.contains_key(&c) {
                return None;
            }
            p = p.map.get(&c).unwrap();
        }

        return Some(p.clone());
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * let obj = Trie::new();
 * obj.insert(word);
 * let ret_2: bool = obj.search(word);
 * let ret_3: bool = obj.starts_with(prefix);
 */

#[cfg(test)]
mod tests {
    use super::Trie;

    #[test]
    fn case_1() {
        let mut t = Trie::new();
        t.insert("andy".to_string());

        let expected = true;
        assert_eq!(expected, t.starts_with("a".to_string()));
        assert_eq!(expected, t.starts_with("and".to_string()));
        assert_eq!(expected, t.starts_with("andy".to_string()));
    }

    #[test]
    fn case_2() {
        let mut t = Trie::new();
        t.insert("apple".to_string());

        assert_eq!(true, t.search("apple".to_string()));
        assert_eq!(false, t.search("app".to_string()));
        assert_eq!(true, t.starts_with("app".to_string()));

        t.insert("app".to_string());
        assert_eq!(true, t.search("app".to_string()));
    }
}

File: python/implement_trie.py
class TrieNode:
    def __init__(self):
        self._char_map: dict[str, TrieNode] = {}
        self._is_terminal: bool = False

    def insert(self, word: str, idx: int) -> None:
        """time complexity: O(n). n is the length of the word."""
        if idx >= len(word):
            self._is_terminal = True
            return
        if word[idx] not in self._char_map:
            self._char_map[word[idx]] = TrieNode()
        self._char_map[word[idx]].insert(word, idx + 1)

    def search(self, word: str, idx: int) -> bool:
        """time complexity: O(n). n is the length of the word."""
        if idx >= len(word):
            return self._is_terminal
        if word[idx] not in self._char_map:
            return False
        return self._char_map[word[idx]].search(word, idx + 1)

    def startsWith(self, prefix: str, idx: int) -> bool:
        """time complexity: O(n). n is the length of the prefix."""
        if idx >= len(prefix):
            return True
        if prefix[idx] not in self._char_map:
            return False
        return self._char_map[prefix[idx]].startsWith(prefix, idx + 1)


class Trie:
    """This is a Trie class."""

    def __init__(self):
        self._root: TrieNode = TrieNode()

    def insert(self, word: str) -> None:
        """Insert a word into the trie."""
        self._root.insert(word, 0)

    def search(self, word: str) -> bool:
        """Returns if the word is in the trie."""
        return self._root.search(word, 0)

    def startsWith(self, prefix: str) -> bool:
        """Returns if there is any word in the trie that starts with the given prefix."""
        return self._root.startsWith(prefix, 0)


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)
def test_trie_case_1():
    """This is a test case for Trie."""
    trie: Trie = Trie()
    trie.insert("apple")
    assert trie.search("apple")  # return True
    assert not trie.search("app")  # return False
    assert trie.startsWith("app")  # return True
    trie.insert("app")
    assert trie.search("app")  # return True


def test_trie_case_2():
    """This is a test case for Trie."""
    trie: Trie = Trie()
    trie.insert("a")
    assert trie.search("a")  # return True
    assert trie.startsWith("a")  # return True
